### PR TITLE
Pawns tracking and range tools fix

### PR DESCRIPTION
**Pull request by kbatbouta**
Edit (Wiri):
- Goal of tracking pawns with the hope of being able to create a data-structure which allows for fast (preferably O(1)  access, 0(n) insertion) access to information about pawns on a given map. Allowing for optimisation of turrets idly looking for targets and pawns searching for cover/targets. 

## Goals 

 Edit (Kbatbouta):

- [x] Rewrite `CanHitFromIgnoringRange` .
- [x] Added `MapStorage.UpdatePawnPos` to implement the previous algorithm.
- [x] Implement a Thing tracker rather a Pawn tracker.
- [x] Optimize the algorithm inner code.
- [x] Implement GC for the tracker.
- [x] Implement startup setup for the tracker.
- [ ] Implement a proper debug tool.
- [ ] Add support for turrets.
- [ ] Add support for a modExtension (maybe).

### CODE REVIEW

**Comment by kbatbouta** [comment]
It's done but it need some more testing.

File: Source/CombatExtended/Storage/MapStorageUtils.cs [hunk 1]
```diff
@@ -0,0 +1,56 @@
+﻿using System;
+using System.Collections.Generic;
+using Verse;
+
+namespace CombatExtended.Storage
+{
+    public static class StorageUtils
+    {
+        public static IEnumerable<Thing> ThingsInRange(this Thing thing, float distance)
```
Comment (simplyWiri) [inline Source/CombatExtended/Storage/MapStorageUtils.cs]: Same here, as discussed prior. We can in some cases just return two integers [startIndex, endIndex], and expose access to the internal array. This means we do not have to do any iteration or return an enumerable.

Usage could be like.
```c#
var pair = thing.ThingsInRange(15);

for(int i = pair.First; i <= pair.Second; i++)
     thing t = thing.Map.CEArray[i];
```
Tuple<int, int> being the returntype, or a named struct which makes the naming more elegant

Also, in the same vein of thought, when returning this enumerable for cases where you wish to order the data in your own way, you can use binary search to find the lower and upper bound. Then return the range of items between the two...
```c#
int lowerBound = BinarySearch(array, thing.posx - distance);
int upperBound = BinarySearch(array, thing.posx + distance);

for(int i = lowerBound; i <= upperBound; i++)
     yield return array[i];
```

Happy to make these changes post-merge. But they should be made.

Also. Why cache x & y, if only x is used then a distance formula to check thoroughly?

**Review by simplyWiri — COMMENTED** [review]
I'd like to see the 'empty' patches gotten rid of and same w/ the duplicates. The algorithms themselves can be changed post merge.

I also don't see a reason for warnings as errors to be disabled

File: Source/CombatExtended/Storage/Harmony/Harmony_Things.cs [hunk 2]
```diff
@@ -50,17 +50,51 @@ public static IEnumerable<CodeInstruction> Transpiler(IEnumerable<CodeInstructio
...
                 return;
-
             thing?.Map?.CEDataStore?.UpdateThingPosition(thing);
         }
     }
 
+    [HarmonyPatch(typeof(Thing), nameof(Thing.Destroy))]
+    public static class Harmony_Thing_Destroy
+    {
+        public static void Prefxi(Thing __instance)
```
[the remark this answers is not in the document]
Comment (kbatbouta) [reply Source/CombatExtended/Storage/Harmony/Harmony_Things.cs]: I know i fixed it. need to push it.

File: Source/CombatExtended/CombatExtended.csproj [hunk 3]
```diff
@@ -14,21 +14,23 @@
...
+    <TreatWarningsAsErrors>false</TreatWarningsAsErrors>
     <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
     <Optimize>true</Optimize>
+    <PlatformTarget>x64</PlatformTarget>
   </PropertyGroup>
 
   <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Release|AnyCPU'">
     <OutputPath>..\..\Assemblies\</OutputPath>
-    <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
+    <TreatWarningsAsErrors>false</TreatWarningsAsErrors>
```
[the remark this answers is not in the document]
Comment (kbatbouta) [reply Source/CombatExtended/CombatExtended.csproj:28]: Ok

File: Source/CombatExtended/Storage/Harmony/Harmony_MapStorage_MapChanged.cs [hunk 4]
```diff
@@ -0,0 +1,26 @@
+﻿using System;
```
[the remark this answers is not in the document]
Comment (kbatbouta) [reply Source/CombatExtended/Storage/Harmony/Harmony_MapStorage_MapChanged.cs]: Still working on them.

File: Source/CombatExtended/Storage/Harmony/Harmony_Things.cs [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+        {
+            if (__instance.isPawn || __instance.isTurret)
+            {
+                if (__instance?.Map == null)
+                    return;
+
+                __instance?.Map?.CEDataStore?.LOC_CACHE_X.Clear();
```
[the remark this answers is not in the document]
Comment (kbatbouta) [reply Source/CombatExtended/Storage/Harmony/Harmony_Things.cs]: For now as a solution for GC.

File: Source/CombatExtended/Storage/Harmony/Harmony_Things.cs [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+                    return;
+                if (__instance.Destroyed || !__instance.Spawned)
+                    return;
+                __instance?.Map?.CEDataStore?.UpdateThingPosition(__instance);
+            }
+        }
+    }
+
+    [HarmonyPatch(typeof(Thing), nameof(Thing.SpawnSetup))]
+    public static class Harmony_Thing_SpawnSetup
```
[the remark this answers is not in the document]
Comment (kbatbouta) [reply Source/CombatExtended/Storage/Harmony/Harmony_Things.cs:102]: Indeed.

File: Source/CombatExtended/Storage/MapStorage.cs [hunk 7]
```diff
@@ -0,0 +1,255 @@
...
+            }
+
+            public static CacheSortable<T> Create(T value, int weight) => new CacheSortable<T>(value, weight);
+        }
+
+        #endregion
+        #region fields       
+
+        public List<CacheSortable<Thing>> LOC_CACHE_X = new List<MapStorage.CacheSortable<Thing>>(100);
+        public List<CacheSortable<Thing>> LOC_CACHE_Z = new List<MapStorage.CacheSortable<Thing>>(100);
```
[the remark this answers is not in the document]
Comment (kbatbouta) [reply Source/CombatExtended/Storage/MapStorage.cs]: I know these are given this style in reminiscing of Python. Sorry.

File: Source/CombatExtended/Storage/MapStorageUtils.cs (hunk 1 above)
Comment (kbatbouta) [reply Source/CombatExtended/Storage/MapStorageUtils.cs]: This allow for O(m) retrieval.

File: Source/CombatExtended/Storage/MapStorage.cs [hunk 8]
```diff
@@ -0,0 +1,255 @@
...
+
+                thing.indexValid = true;
+            }
+        }
+
+        #endregion
+        #region static_methods
+
+        public static float AVG_INSERTION_TIME = 0;
+        public static void InsertionSort<T>(
```
[the remark this answers is not in the document]
Comment (kbatbouta) [reply Source/CombatExtended/Storage/MapStorage.cs]: This is not for insertion. it's for already inserted values that changed. I could make it a binary insertion in the method above it.

File: Source/CombatExtended/Storage/MapStorageUtils.cs (hunk 1 above)
Comment (kbatbouta) [reply Source/CombatExtended/Storage/MapStorageUtils.cs]: These allow me for updating the value in average of O(m) where m ~= 1

**Pull request merged**